File: Sources/LoeBalance/Core/AppError.swift
import Foundation
import Security

enum AppError: Error, Equatable, Sendable {
    case invalidURL
    case transport(URLError)
    case serverStatus(Int)
    case apiEnvelope(code: Int, message: String)
    case unauthorized
    case rateLimited(retryAfter: Date?)
    case invalidResponse
    case keychainStatus(OSStatus)
    case loginChallenge
}

File: windows/src/LoeBalance.Core/Models/AppError.cs
namespace LoeBalance.Core.Models;

public enum AppErrorKind
{
    Transport,
    Unauthorized,
    RateLimited,
    ServerStatus,
    ApiEnvelope,
    InvalidResponse,
    InvalidUrl,
    CredentialStore
}

public sealed class AppException : Exception
{
    public AppErrorKind Kind { get; }
    public int? StatusCode { get; }
    public DateTimeOffset? RetryAfter { get; }

    public AppException(
        AppErrorKind kind,
        string message,
        Exception? innerException = null,
        int? statusCode = null,
        DateTimeOffset? retryAfter = null)
        : base(message, innerException)
    {
        Kind = kind;
        StatusCode = statusCode;
        RetryAfter = retryAfter;
    }
}
